text/unification.md

```diff
@@ -0,0 +1,268 @@
+
+  We call projection of variables the subcase when `u` is a variable and `E` is empty. In practice, Coq implements only projections of variables and projections of terms of the form `C t₁ ... tₙ` for `C` a constructor, with `E` empty (but PR [#14798](https://github.com/coq/coq/pull/14798) extends this to `E` made of nesting of `match` and projections).
+
+- *Imitation*, that is, imitating the structure of `t`, or part of it, and generating subproblems about the subterms of the imitated part of `t`. 
+
+  For instance, if `C` is a constructor, the equation `Γ ⊢ ?e@{subst} =?= C t₁ ... tₙ` reduces to the collection of equations `Γ ⊢ ?eᵢ@{subst} =?= tᵢ` by declaring new existential variables `?eᵢ` of appropriate type in the same context as `?e` and by setting `?e := C ?e₁@{dom(Δ)} ... ?eₙ@{dom(Δ)}`. 
+
+  Similarly, if the head symbol of `t` is a binder, new existential variables are generated for the subterms of the binder in a context extended with the new binder.
+
+## Efficiency non-determinism
+
+The non-determinism comes from the potentially different ways to project and from the ability to also sometimes imitate. For instance, there are 2 possible projections and one possible imitation to resolve `Γ ⊢ ?e@{a:=t,b:=t} =?= t` for `t` variable-free.
+
+There is also a form of non-determinism which comes from reducing unification problems or not in normal form:
+- When resolving `Γ ⊢ c t =?= c' t'` with `c` and `c'` distinct unfoldable constants, there is a choice in reducing `c` or `c'` in the hope that the problem is eventually found satisfiable. This choice can lead to arbitrarily long computations so, controlling unfolding is important (there are similar problems when efficiently comparing the convertibility of two terms).
+- It is also useful to limit unfolding of constants so as to obtain solutions as close as possible to the original form of the terms involved.
+
+## Higher order pattern unification
+
+*Higher order pattern unification* (see Miller 1991 and Pfenning 1991) is a special case of simple unification problem when the substitution is made only of distinct variables. In this case, there is a single solution obtained by projecting the unique possible variable or imitating if not a variable.
+
+Pattern unification has various refinements (see e.g. Libal and Miller 2016) such as:
+- expecting distinct variables only among the variables that actually occur in the term to match
+- identifying arguments that are irrefutable patterns over variables at their leaves (e.g. `(x,(y,z))` with those variables (i.e. a form of curryfication) (see PR [#14798](https://github.com/coq/coq/pull/14798))
```

Comment:
This is only guaranteed to generate a good solution either (a) in the presence of eta conversion for the record type, or (b) when no subterm of the constructor tree, other than the leaves, occurs in the RHS.  e.g., `x, y, z |- ?e@{x:=(x,(y,z))} == (y,z)` can be solved with either `snd x` or `(fst (snd x), snd (snd x))`.

Reply:
Right.